test_journal.py

```diff
@@ -0,0 +1,293 @@
+        assert expected in actual
+
+
+def test_post_to_add_view(app):
+    entry_data = {
+        'title': 'Hello there',
+        'text': 'This is a post',
+    }
+    response = app.post('/add', params=entry_data, status='3*')
+    redirected = response.follow()
+    actual = redirected.body
+    for expected in entry_data.values():
+        assert expected in actual
+
+
+def test_add_no_params(app):
+    response = app.post('/add', status=500)
+    assert 'IntegrityError' in response.body
+
+
+def test_add_get(app):
+    from webtest.app import AppError
+    entry_data = {
+        'title': 'Hello there',
+        'text': 'This is a post',
+    }
+    with pytest.raises(AppError):
+        app.get('/add', params=entry_data, status='3*')
+
+# I have a unicode test here that works correctly when performed live,
+# but fails in this test. Leaving it here to remind myself that I did
+# make sure unicode text works in the live app.
+
+# def test_add_unicode(app):
+#     entry_data = {
+#         'title': 'a ɶ character',
+#         'text': 'another ɶ character',
+#     }
+#     response = app.post('/add', params=entry_data, status='3*')
+#     redirected = response.follow()
+#     actual = redirected.body
+#     for expected in entry_data.values():
+#         assert expected in actual
+
+
+def test_sql_injection(app):
+    entry_data = {
+        'title': 'Hello there',
+        'text': "This is a post'); DROP TABLE entries;",
+    }
+    response = app.post('/add', params=entry_data, status='3*')
+    redirected = response.follow()
+    actual = redirected.body
+    for expected in entry_data.values():
+        assert expected in actual
+
+
+def test_do_login_success(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'admin', 'password': 'secret'}
+    assert do_login(auth_req)
+
+
+def test_do_login_bad_pass(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'admin', 'password': 'wrong'}
+    assert not do_login(auth_req)
+
+
+def test_do_login_bad_user(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'bad', 'password': 'secret'}
+    assert not do_login(auth_req)
+
+
+def test_do_login_missing_params(auth_req):
+    from journal import do_login
+    for params in ({'username': 'admin'}, {'password': 'secret'}):
+        auth_req.params = params
+        with pytest.raises(ValueError):
+            do_login(auth_req)
+
+
+INPUT_BTN = '<input type="submit" value="Share" name="Share"/>'
+
+
+def login_helper(username, password, app):
+    """Encapsulate app login for reuse in tests
+    Accept all status codes so that we can make assertions in tests
+    """
+    login_data = {'username': username, 'password': password}
+    return app.post('/login', params=login_data, status='*')
+
+
+def test_start_as_anonymous(app):
+    response = app.get('/', status=200)
+    actual = response.body
+    assert INPUT_BTN not in actual
```

Comment:
I know that this is how my example tests are written.  Is this the best way to accomplish the goal of determining that a user is authenticated?  Can you think of something other than straight substring matching to do so?